controllers/routing/routing_reconcile_resources.go

```diff
@@ -117,6 +121,56 @@ func propagateHostsToWatchedCR(target *unstructured.Unstructured, data spi.Routi
 	return nil
 }
 
+func (r *PlatformRoutingController) patchResourceMetadata(ctx context.Context, req ctrl.Request, sourceRes *unstructured.Unstructured) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentRes := &unstructured.Unstructured{}
+		currentRes.SetGroupVersionKind(sourceRes.GroupVersionKind())
+
+		if err := r.Client.Get(ctx, req.NamespacedName, currentRes); err != nil {
+			return fmt.Errorf("failed re-fetching resource: %w", err)
+		}
+
+		patch := client.MergeFrom(currentRes)
+		if errPatch := r.Client.Patch(ctx, sourceRes, patch); errPatch != nil {
+			return fmt.Errorf("failed to patch resource with updated annotations: %w", errPatch)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to patch resource metadata with retry: %w", err)
+	}
+
+	return nil
+}
+
+func (r *PlatformRoutingController) addResourceFinalizer(ctx context.Context, req ctrl.Request, sourceRes *unstructured.Unstructured) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentRes := &unstructured.Unstructured{}
+		currentRes.SetGroupVersionKind(sourceRes.GroupVersionKind())
+
+		if err := r.Client.Get(ctx, req.NamespacedName, sourceRes); err != nil {
+			return fmt.Errorf("failed re-fetching resource: %w", err)
+		}
+
+		if controllerutil.AddFinalizer(sourceRes, finalizerName) {
+			patch := client.MergeFrom(currentRes)
+			if err := r.Client.Patch(ctx, sourceRes, patch); err != nil {
+				return fmt.Errorf("failed to patch resource with finalizer: %w", err)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed adding finalizer with retry: %w", err)
+	}
+
+	return nil
+}
+
```

Comment:
We can reduce duplication here.:

```suggestion
func (r *PlatformRoutingController) patch(ctx context.Context, target *unstructured.Unstructured) error {
	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
		currentRes := &unstructured.Unstructured{}
		currentRes.SetGroupVersionKind(sourceRes.GroupVersionKind())

		if err := r.Client.Get(ctx, client.ObjectKeyFromObject(target), currentRes); err != nil {
			return fmt.Errorf("failed re-fetching resource: %w", err)
		}

		patch := client.MergeFrom(currentRes)
		if errPatch := r.Client.Patch(ctx, target, patch); errPatch != nil {
			return fmt.Errorf("failed to patch: %w", errPatch)
		}

		return nil
	})

	if err != nil {
		return fmt.Errorf("failed to patch resource metadata with retry: %w", err)
	}

	return nil
}

func (r *PlatformRoutingController) addFinalizer(ctx context.Context, target *unstructured.Unstructured) error {
	if controllerutil.AddFinalizer(target, finalizerName) {
		return r.patch(ctx, target)
	}

	return nil
}
```

Reply:
Another thing I wonder, as we do not operate on `PartialObjectMetadata` (yet?), is to trim the object to build a patch with:

```golang
unstructured.RemoveNestedField(target.Object, "spec")
unstructured.RemoveNestedField(target.Object, "status")
```

WDYT @aslakknutsen @cam-garrison?

Reply:
In addition,  we can treat `patchResourceMetadata` as a generic `patch`  operation and move it to `pkg/cluster/unstruct.go`, so this could look like this:

```golang

type MutateUnstr func(target *unstructured.Unstructured)

func Patch(ctx context.Context, cli client.Client, target *unstructured.Unstructured, mutFns...MutateUnstrFn) error {
	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
		currentRes := &unstructured.Unstructured{}
		currentRes.SetGroupVersionKind(target.GroupVersionKind())

		if err := cli.Get(ctx, client.ObjectKeyFromObject(target), currentRes); err != nil {
			return fmt.Errorf("failed re-fetching resource: %w", err)
		}

		for _, fn := range mutFns {
			fn(target)
		}

		patch := client.MergeFrom(currentRes)
		if errPatch := cli.Patch(ctx, target, patch); errPatch != nil {
			return fmt.Errorf("failed to patch: %w", errPatch)
		}

		return nil
	})

	if err != nil {
		return fmt.Errorf("failed to patch resource metadata with retry: %w", err)
	}

	return nil
}
```

and the caller in the main reconcile could do this:

```golang
errs = append(errs, Patch(ctx, r.Client, sourceRes, func(target *unstructured.Unstructured) {
	unstructured.RemoveNestedField(target.Object, "spec")
	unstructured.RemoveNestedField(target.Object, "status")
}))
```

Reply:
as long as the patch doesn't remove those...?

Reply:
> as long as the patch doesn't remove those...?

I think I confused the behaviour with the `RawPatch`. In this case, it will track the changes and assumes both spec and status should be removed.

Lets drop the idea of `MutateFunc` but promote the Patch to top-level func.